StackTask.playground/Contents.swift

```diff
@@ -1,60 +1,64 @@
+/// The protocol defines the necessary functionality that any stack should provide.
 protocol StackOperations {
-    func push (element: Int)
-    func pop() -> Int?
-    func top() -> Int?
+    associatedtype Element
+    func push(element: Element)
+    func pop() -> Element?
+    func top() -> Element?
 }
 
-class Stack: StackOperations {
-    private var stackArray: [Int] = []
-
-    // Adds element on the top of the stack.
-    func push(element: Int) {
+/// The class is based on basic principle of last-in-first-out (LIFO).
+class Stack<Element>: StackOperations {
+    private var stackArray: [Element] = []
+    
+    /// Adds element on the top of the stack.
+    /// - Parameter element: The element to add to the stack.
+    /// - Complexity: O(1)
+    func push(element: Element) {
         stackArray.append(element)
     }
     
-    // Function removes and returns the top element of the stack. Returns nil when the invoking stack is empty.
-    func pop() -> Int? {
-        if stackArray.isEmpty {
-            return nil
-        } else {
-            return stackArray.removeLast()
-        }
+    /// Removes and returns the top element of the stack.
+    /// - Returns: The top element of the stack, if stack is not empty. Otherwise: nil.
+    /// - Complexity: O(1)
+    func pop() -> Element? {
+        stackArray.popLast()
     }
     
-    // Function returns the top element of the stack, but does not remove it. Returns nil when the invoking stack is empty.
-    func top() -> Int? {
+    /// Returns the top element of the stack.
+    /// - Returns: The top element of the stack, if stack is not empty. Otherwise: nil.
+    /// - Complexity: O(1)
+    func top() -> Element? {
         return stackArray.last
     }
 }
 
-class StackStatistics: Stack {
-    private var minimalElements: [Int] = []
+/// The class is based on basic principle of last-in-first-out (LIFO).
+class StackStatistics<Element: Comparable>: Stack<Element> {
+    private var minimalElements: [Element] = []
 
-    // Function adds element on the top of the steck and adds this element into minimal element array, if new element is less than the current.
-    // Implements basic function.
-    override func push(element: Int) {
-        if minimalElements.isEmpty || element <= minimalElements.last! {
+    override func push(element: Element) {
+        if let minElement = minimalElements.last {
+            minimalElements.append(min(element, minElement))
+        } else {
             minimalElements.append(element)
         }
         super.push(element: element)
     }
```

Comment:
nit:

smth like:

```suggestion
        minimalElements.append(min(element, minimalElements.last ?? element))
        super.push(element: element)
    }
```

?

Reply:
Thank u =) it looks much better. I fixed!